In [8]:
import keras
from keras import layers, models

In [9]:
model=models.Sequential()

model.add(layers.Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1) ))
model.add(layers.MaxPooling2D((2,2)) )

model.add(layers.Conv2D(64, (3,3), activation='relu'))
model.add(layers.MaxPooling2D((2,2)) )

model.add(layers.Conv2D(128, (3,3), activation='relu'))


In [10]:
model.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_13 (Conv2D)           (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_9 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_14 (Conv2D)           (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_10 (MaxPooling (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_15 (Conv2D)           (None, 3, 3, 128)         73856     
Total params: 92,672
Trainable params: 92,672
Non-trainable params: 0
_________________________________________________________________


1(32), 2(64), 3(64)

1x(3x3x32)+(32) = 320 -> conv2d_1 param  
32x(3x3x64)+(64) = 18496 -> conv2d_2 param  
64x(3x3x64)+64 = 36928 ->conv2d_3 param  

In [11]:
model.add(layers.Flatten())
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

In [12]:
model.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_13 (Conv2D)           (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_9 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_14 (Conv2D)           (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_10 (MaxPooling (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_15 (Conv2D)           (None, 3, 3, 128)         73856     
_________________________________________________________________
flatten_1 (Flatten)          (None, 1152)              0         
_________________________________________________________________
dense_1 (Dense)              (None, 128)              

conv1(32), conv2(64), conv3(64), flatten(128), dense(128), dense(10)

### 1x(3x3x[32])+[32] = 320 -> Conv2d_1 param  
### [32]x(3x3x[[64]])+[[64]] = 18496 -> Conv2d_2 param  
### [[64]]x(3x3x<128>)+<128> = 73586 -> Conv2d_3 param  
### <128> x 9 = [(1152)] -> Flatten
### [(1152)] x (128) +(128) = -> dense_1  
### (128)x10+10 -> dense_2  

In [13]:
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()


In [14]:
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [15]:
model.compile(optimizer='RMSprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5, batch_size=64)

Instructions for updating:
Use tf.cast instead.
Epoch 1/5
60000/60000 [==============================] - 38s 628us/step - loss: 0.1517 - accuracy: 0.9519
Epoch 2/5
60000/60000 [==============================] - 36s 598us/step - loss: 0.0411 - accuracy: 0.9876
Epoch 3/5
60000/60000 [==============================] - 37s 615us/step - loss: 0.0278 - accuracy: 0.9914
Epoch 4/5
60000/60000 [==============================] - 35s 578us/step - loss: 0.0212 - accuracy: 0.9937
Epoch 5/5
60000/60000 [==============================] - 35s 581us/step - loss: 0.0175 - accuracy: 0.9945


In [ ]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

In [ ]:
test_acc

# #####################

In [ ]:
model=models.Sequential()
model.add(layers.Conv2D(10, (3,3), activation='relu', input_shape=(39,39,3) ))
model.add(layers.Conv2D(20, (5,5), strides=2, activation='relu'))
model.add(layers.Conv2D(40, (5,5), strides=2, activation='relu'))

model.summary()

data = 1,000
batch = 200

1 step(batch) = 200 
5 step(batch) = 1,000 = 1 epoch
1 batch 당 가중치를 수정.